qos/evm/endpoint_selection.go

```diff
@@ -19,80 +21,170 @@ var (
 
 	if len(availableEndpoints) == 0 {
-		return nil, errEmptyEndpointListObs
+		return nil, nil, errEmptyEndpointListObs
 	}
 
 	logger.Info().Msgf("About to filter through %d available endpoints", len(availableEndpoints))
 
+	var filteredEndpointsAddr protocol.EndpointAddrList
+	var validationResults []*qosobservations.EndpointValidationResult
+
 	// TODO_FUTURE: use service-specific metrics to add an endpoint ranking method
 	// which can be used to assign a rank/score to a valid endpoint to guide endpoint selection.
-	var filteredEndpointsAddr protocol.EndpointAddrList
 	for _, availableEndpointAddr := range availableEndpoints {
 		logger := logger.With("endpoint_addr", availableEndpointAddr)
 		logger.Info().Msg("processing endpoint")
 
 		endpoint, found := ss.endpointStore.endpoints[availableEndpointAddr]
 		if !found {
 			logger.Error().Msgf("❓ SKIPPING endpoint %s because it was not found in PATH's endpoint store.", availableEndpointAddr)
+
+			// Create validation result for endpoint not found
+			failureDetails := "endpoint not found in PATH's endpoint store"
+			result := &qosobservations.EndpointValidationResult{
+				EndpointAddr:   string(availableEndpointAddr),
+				Success:        false,
+				FailureReason:  qosobservations.EndpointValidationFailureReason_ENDPOINT_VALIDATION_FAILURE_REASON_ENDPOINT_NOT_FOUND.Enum(),
+				FailureDetails: &failureDetails,
+			}
+			validationResults = append(validationResults, result)
 			continue
 		}
 
 		if err := ss.basicEndpointValidation(endpoint); err != nil {
 			logger.Error().Err(err).Msgf("❌ SKIPPING %s endpoint because it failed basic validation: %v", availableEndpointAddr, err)
+
+			// Create validation result for validation failure
+			failureReason := ss.categorizeValidationFailure(err)
+			errorMsg := err.Error()
+			result := &qosobservations.EndpointValidationResult{
+				EndpointAddr:   string(availableEndpointAddr),
+				Success:        false,
+				FailureReason:  &failureReason,
+				FailureDetails: &errorMsg,
+			}
+			validationResults = append(validationResults, result)
 			continue
 		}
 
+		// Endpoint passed validation - record success and add to valid list
+		result := &qosobservations.EndpointValidationResult{
+			EndpointAddr: string(availableEndpointAddr),
+			Success:      true,
+			// FailureReason and FailureDetails are nil for successful validations
+		}
+		validationResults = append(validationResults, result)
 		filteredEndpointsAddr = append(filteredEndpointsAddr, availableEndpointAddr)
 		logger.Info().Msgf("✅ endpoint %s passed validation", availableEndpointAddr)
 	}
 
-	return filteredEndpointsAddr, nil
+	return filteredEndpointsAddr, validationResults, nil
+}
+
+// categorizeValidationFailure determines the failure reason category based on the validation error.
+func (ss *serviceState) categorizeValidationFailure(err error) qosobservations.EndpointValidationFailureReason {
```

Comment:
We're not using `ss.` at all.

Can you move this into a different place/file/package where we do error validation?

I don't think this is evm specific.